### PR TITLE
[JENKINS-47793] - Jenkins master and agent configuration pages do not verify negative executor numbers

### PR DESCRIPTION
**Pull request by ksenia-nenasheva**
See [JENKINS-47793](https://issues.jenkins-ci.org/browse/JENKINS-47793).

Field validation for "# of executors" did not work correctly in agent/master configuration pages. It was showing a page with an error and log on "Save". For agents it was also possible to set a negative number in the "# of executors" field.

After this fix:
For the master, "# of executors" should be a non-negative number (include 0).
For agents, "# of executors" should be a positive number (excluding 0).

### Proposed changelog entries

* Entry 1: Fix handling and field validation of "# of executors" in master and agent configurations.
* ...

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers

File: core/src/main/java/hudson/model/Computer.java [hunk 1]
```diff
@@ -1467,6 +1467,11 @@ public void doConfigSubmit( StaplerRequest req, StaplerResponse rsp ) throws IOE
             throw new FormException(Messages.ComputerSet_SlaveAlreadyExists(proposedName), "name");
         }
 
+        String nExecutors = req.getSubmittedForm().getString("numExecutors");
+        if (!nExecutors.matches("^(\\d*[1-9]\\d*)$")) {
```
Comment (jglick) [inline core/src/main/java/hudson/model/Computer.java]: `^` and `$` are gratuitous when using `matches` (as opposed to, say, `Matcher.find`). So are the `()`, since you are not asking for any `group`.

More to the point, this will improperly reject `0` AFAICT, which is a valid executor number (for the master node anyway). Or is this code called only for agents?

Maybe simpler to just `Integer.parseInt`, catch `NumberFormatException`, and reject a negative result.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 2]
```diff
@@ -2758,7 +2759,7 @@ public InitMilestone getInitLevel() {
         return initLevel;
     }
 
-    public void setNumExecutors(int n) throws IOException {
+    public void setNumExecutors(@Nonnegative int n) throws IOException {
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: No runtime effect AFAIK; should really be matched by an `IllegalArgumentException`.

File: core/src/main/resources/jenkins/model/Messages.properties [hunk 3]
```diff
@@ -26,6 +26,7 @@
 Hudson.BadPortNumber=Bad port number {0}
 Hudson.Computer.Caption=Master
 Hudson.Computer.DisplayName=master
+Hudson.Computer.IncorrectNumberOfExecutors=Incorrect field "# of executors". It should be no negative number.
```
Comment (jglick) [inline core/src/main/resources/jenkins/model/Messages.properties]: `non`

File: war/src/main/webapp/scripts/hudson-behavior.js [hunk 4]
```diff
@@ -694,9 +694,12 @@ var jenkinsRules = {
     "INPUT.required" : function(e) { registerRegexpValidator(e,/./,"Field is required"); },
 
 // validate form values to be an integer
-    "INPUT.number" : function(e) { registerRegexpValidator(e,/^(\d+|)$/,"Not an integer"); },
+    "INPUT.number" : function(e) { registerRegexpValidator(e,/^\-?(\d+)$/,"Not an integer"); },
+    "INPUT.non-negative-number" : function(e) {
+        registerRegexpValidator(e,/^(\d+)$/,"Not a non-negative number");
```
Comment (jglick) [inline war/src/main/webapp/scripts/hudson-behavior.js]: gratuitous `()` BTW

**Comment by oleg-nenashev** [comment]
WiO, I'd guess

**Comment by ksenia-nenasheva** [comment]
@jglick Done.
Yes, this function is only for agents.
In this PR the master may have a non-negative number of executors (positive and zero), agents may have only a positive number of executors.

Thanks for the review!

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers back to review

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 5]
```diff
@@ -2758,7 +2759,7 @@ public InitMilestone getInitLevel() {
         return initLevel;
     }
 
-    public void setNumExecutors(int n) throws IOException {
+    public void setNumExecutors(@Nonnegative int n) throws IOException, IllegalArgumentException {
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/Jenkins.java]: Hmm, does it really throw `IllegalArgumentException`?

**Review by oleg-nenashev — COMMENTED** [review]
It seems that the error processing is not complete. The IllegalArgumentException exceptioneither needs to be thrown or removed from `throws` (I would vote for the first option). The rest LGTM

**Review by oleg-nenashev — APPROVED** [review]

File: core/src/main/java/jenkins/model/MasterBuildConfiguration.java [hunk 6]
```diff
@@ -51,6 +51,11 @@ public boolean configure(StaplerRequest req, JSONObject json) throws FormExcepti
         Jenkins j = Jenkins.getInstance();
         try {
             // for compatibility reasons, this value is stored in Jenkins
```
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/MasterBuildConfiguration.java]: comment displaced, refers to `j.setNumExecutors` line.

**Review by daniel-beck — APPROVED** [review]
LGTM expect for the displaced comment.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 7]
```diff
@@ -2758,16 +2759,23 @@ public InitMilestone getInitLevel() {
...
         if (this.numExecutors != n) {
             this.numExecutors = n;
             updateComputerList();
             save();
         }
     }
-
-
-
+
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: BTW it is best to revert unrelated whitespace changes in PRs.

File: core/src/main/java/jenkins/model/MasterBuildConfiguration.java [hunk 8]
```diff
@@ -51,6 +51,11 @@ public boolean configure(StaplerRequest req, JSONObject json) throws FormExcepti
...
+            String num = json.getString("numExecutors");
+            if (!num.matches("^(\\d+)")) {
```
Comment (jglick) [inline core/src/main/java/jenkins/model/MasterBuildConfiguration.java]: BTW `^` and `(`…`)` both gratuitous here; could be simplified to

```java
if (!num.matches("\\d+")) {
```

File: core/src/main/java/hudson/model/Computer.java [hunk 9]
```diff
@@ -1467,6 +1467,11 @@ public void doConfigSubmit( StaplerRequest req, StaplerResponse rsp ) throws IOE
             throw new FormException(Messages.ComputerSet_SlaveAlreadyExists(proposedName), "name");
         }
 
+        String nExecutors = req.getSubmittedForm().getString("numExecutors");
+        if (Integer.parseInt(nExecutors)<=0) {
+            throw new FormException(Messages.Slave_InvalidConfig_Executors(nodeName), "numExecutors");
```
Comment (jglick) [inline core/src/main/java/hudson/model/Computer.java]: As below, this would more naturally live in `Slave.setNumExecutors` but that would be a source-incompatible change.

File: core/src/main/java/jenkins/model/MasterBuildConfiguration.java [hunk 10]
```diff
@@ -51,6 +51,11 @@ public boolean configure(StaplerRequest req, JSONObject json) throws FormExcepti
...
+                throw new FormException(Messages.Hudson_Computer_IncorrectNumberOfExecutors(),"numExecutors");
```
Comment (jglick) [inline core/src/main/java/jenkins/model/MasterBuildConfiguration.java]: It would be nicer to just throw the `FormException` from the setter, so you do not need to duplicate code, but I suppose the problem is that this is a checked exception.

File: war/src/main/webapp/scripts/hudson-behavior.js [hunk 11]
```diff
@@ -694,9 +694,12 @@ var jenkinsRules = {
...
 
 // validate form values to be an integer
-    "INPUT.number" : function(e) { registerRegexpValidator(e,/^(\d+|)$/,"Not an integer"); },
+    "INPUT.number" : function(e) { registerRegexpValidator(e,/^\-?(\d+)$/,"Not an integer"); },
+    "INPUT.non-negative-number" : function(e) {
+        registerRegexpValidator(e,/^\d+$/,"Not a non-negative number");
+    },
     "INPUT.positive-number" : function(e) {
-        registerRegexpValidator(e,/^(\d*[1-9]\d*|)$/,"Not a positive integer");
+        registerRegexpValidator(e,/^(\d*[1-9]\d*)$/,"Not a positive integer");
```
Comment (jglick) [inline war/src/main/webapp/scripts/hudson-behavior.js]: `/[1-9]\d*/` might be a more intuitive regexp FWIW.

**Review by jglick — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
Needs re-review by @jglick, I'd guess

File: war/src/main/webapp/scripts/hudson-behavior.js [hunk 12]
```diff
@@ -694,9 +694,12 @@ var jenkinsRules = {
...
 
 // validate form values to be an integer
-    "INPUT.number" : function(e) { registerRegexpValidator(e,/^(\d+|)$/,"Not an integer"); },
+    "INPUT.number" : function(e) { registerRegexpValidator(e,/^\-?(\d+)$/,"Not an integer"); },
+    "INPUT.non-negative-number" : function(e) {
+        registerRegexpValidator(e,/^\d+$/,"Not a non-negative number");
+    },
     "INPUT.positive-number" : function(e) {
-        registerRegexpValidator(e,/^(\d*[1-9]\d*|)$/,"Not a positive integer");
+        registerRegexpValidator(e,/[1-9]\d*/,"Not a positive integer");
```
Comment (jglick) [inline war/src/main/webapp/scripts/hudson-behavior.js]: This needs to be anchored with `^…$`, as it seems JavaScript `String.match(RegExp)` searches for the first match, like Java’s `find` but unlike Java’s `matches`.

**Review by jglick — CHANGES_REQUESTED** [review]

**Review by jglick — APPROVED** [review]

**Comment by daniel-beck** [comment]
@oleg-nenashev Why is this on-hold?

**Comment by oleg-nenashev** [comment]
It can be shipped now. Iirc it was just because of minor fixes after the
previous @jglick's approval
